components/sup/tests/utils/test_sup.rs

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Encapsulate running the `hab-sup` executable for tests.
+
+use std::env;
+use std::process::{Child, Command, Stdio};
+use std::path::{Path, PathBuf};
+use std::string::ToString;
+
+use super::test_butterfly;
+
+extern crate rand;
+use self::rand::distributions::{IndependentSample, Range};
+
+pub struct TestSup {
+    pub hab_root: PathBuf,
+    pub origin_name: String,
+    pub package_name: String,
+    pub service_group: String,
+    pub http_port: u16,
+    pub butterfly_port: u16,
+    pub butterfly_client: test_butterfly::Client,
+    pub cmd: Command,
+    pub process: Option<Child>,
+}
+
+/// Return a random unprivileged TCP port number.
+fn random_port() -> u16 {
+    let between = Range::new(1024, ::std::u16::MAX);
+    let mut rng = rand::thread_rng();
+    between.ind_sample(&mut rng)
+}
+
+/// Find an executable relative to the current integration testing
+/// executable.
+///
+/// Thus if the current executable is
+///
+///    /home/me/habitat/target/debug/deps/compilation-ccaf2f45c24e3840
+///
+/// and we look for `hab-sup`, we'll find it at
+///
+///    /home/me/habitat/target/debug/hab-sup
+///
+fn find_exe<B>(binary_name: B) -> PathBuf
+    where B: AsRef<Path>
+{
+    let exe_root = env::current_exe()
+        .unwrap()
+        .parent()  // deps
+        .unwrap()
+        .parent()  // debug
+        .unwrap()
+        .to_path_buf();
+    let bin = exe_root.join(binary_name.as_ref());
+    assert!(
+        bin.exists(),
+        format!("Expected to find a {:?} executable at {:?}", binary_name.as_ref(), bin)
+    );
+    bin
+}
+
+impl TestSup {
+    /// Create a new `TestSup` that will listen on randomly-selected
+    /// ports for both gossip and HTTP requests so tests run in
+    /// parallel don't step on each other.
+    ///
+    /// See also `new`.
+    pub fn new_with_random_ports<R, O, P, S>(
```

Comment:
Ah yep, that's part of the nightly API and I didn't even see the header! `AtomicIsize` is available to us and should do the job just fine.

I'm cool with however you want to avoid the port conflict - a retry is just fine!